### DISCUSSION TITLE
Handle missing sources - possible improvements

### DISCUSSION THREAD

**Discussion by kpodsiad** [scalameta/metals#3409] [General]
https://github.com/scalameta/metals/issues/3397 made me think.
I think we should do more in situations when source files are missing. I thought about:
- warn users about missing sources more directly, for example when someone tries 'go to definition' we could show a warning or an error 'there are no sources to show'. What @ayoub-benali did in https://github.com/scalameta/metals/pull/3401 is great, but we can do even more.
- using https://github.com/scalameta/metals/pull/3247 to show decompiled sources with some annotation 'you are viewing decompiled sources'. It's my preferred option because it already provides a solution to the problem.

I know it's not a common scenario that sources are missing, but anything is better than 'nothing happens'. What do you think?

**Comment by Arthurm1** [comment]
I've been having a look at the 2nd part of that.  Not sure whether I'll get it done that soon though as it's an area of code I know little about.

See [discord](https://discord.com/channels/632642981228314653/718047292749381662/925034198878806066) for some discussion

**Comment by tgodzik** [comment]
For the first part, I think that would be also pretty useful, but  maybe we should show it the same way we show misconfigured targets?

**Comment by Arthurm1** [comment]
https://github.com/scalameta/metals/pull/3380 does show if sources exist for jars but you do have to run it - it doesn't make it obvious if someone's trying to navigate to the source.

**Comment by kpodsiad** [comment]
@Arthurm1 Happy to hear that you thought about this idea too and you're even working on it.

> #3380 does show if sources exist for jars but you do have to run it - it doesn't make it obvious if someone's trying to navigate to the source.

Exactly, responding to the user's action should be our main goal. Everything else is just an addition to it.

**Reply by tgodzik** [reply]
That might be hard to do, since we don't have the src.zip indexed and we don't know what definitions should use it.

**Comment by Arthurm1** [comment]
I can't take credit - I was looking into it because of [this](https://github.com/scalameta/metals-feature-requests/issues/210)
